starfish/codebook.py

```diff
@@ -0,0 +1,545 @@
+    code_length()
+        return the total length of the codes in the codebook
+
+    Attributes
+    ----------
+    Constants.CODEWORD
+        name of codeword field in spaceTx spec
+    Constants.GENE
+        name of gene specifier in spaceTx spec
+    Constants.VALUE
+        name of value specifier in SpaceTx spec
+
+    Examples
+    --------
+
+    >>> from starfish.util.synthesize import SyntheticData
+    >>> sd = SyntheticData(n_ch=3, n_hyb=4, n_codes=2)
+    >>> sd.codebook()
+    <xarray.Codebook (gene_name: 2, c: 3, h: 4)>
+    array([[[0, 0, 0, 0],
+            [0, 0, 1, 1],
+            [1, 1, 0, 0]],
+
+           [[1, 0, 0, 0],
+            [0, 0, 1, 0],
+            [0, 1, 0, 1]]], dtype=uint8)
+    Coordinates:
+      * gene_name  (gene_name) object 08b1a822-a1b4-4e06-81ea-8a4bd2b004a9 ...
+      * c          (c) int64 0 1 2
+      * h          (h) int64 0 1 2 3
+
+    See Also
+    --------
+    TODO <link to spaceTx format>
+
+    """
+
+    class Constants(AugmentedEnum):
+        CODEWORD = 'codeword'
+        GENE = 'gene_name'
+        VALUE = 'v'
+
+    @property
+    def code_length(self) -> int:
+        """return the length of codes in this codebook"""
+        return int(np.dot(*self.shape[1:]))
+
+    @classmethod
+    def _empty_codebook(cls, code_names: Sequence[str], n_ch: int, n_hyb: int):
+        """create an empty codebook of shape (code_names, n_ch, n_hyb)
+
+        Parameters
+        ----------
+        code_names : Sequence[str]
+            the genes to be coded
+        n_ch : int
+            number of channels used to build the codes
+        n_hyb : int
+            number of hybridization rounds used to build the codes
+
+        Examples
+        --------
+        >>> from starfish.codebook import Codebook
+        >>> Codebook._empty_codebook(['ACTA', 'ACTB'], n_ch=3, n_hyb=2)
+        <xarray.Codebook (gene_name: 2, c: 3, h: 2)>
+        array([[[0, 0],
+                [0, 0],
+                [0, 0]],
+
+               [[0, 0],
+                [0, 0],
+                [0, 0]]], dtype=uint8)
+        Coordinates:
+          * gene_name  (gene_name) object 'ACTA' 'ACTB'
+          * c          (c) int64 0 1 2
+          * h          (h) int64 0 1
+
+        Returns
+        -------
+        Codebook :
+            codebook whose values are all zero
+
+        """
+        codes_index = pd.Index(code_names, name=Codebook.Constants.GENE.value)
+        return cls(
+            data=np.zeros((codes_index.shape[0], n_ch, n_hyb), dtype=np.uint8),
+            coords=(
+                codes_index,
+                pd.Index(np.arange(n_ch), name=Indices.CH.value),
+                pd.Index(np.arange(n_hyb), name=Indices.HYB.value),
+            )
+        )
+
+    @classmethod
+    def from_code_array(
+            cls, code_array: List[Dict[str, Any]],
+            n_hyb: Optional[int]=None, n_ch: Optional[int]=None) -> "Codebook":
```

Comment:
Is there any compelling reason to ever have n_hyb / n_ch be optional?

Reply:
There are many cases where you can infer them by simply taking max(hyb) and max(ch). I think most real-world use cases will be like this, so I wanted to leave this in. 

For my tests where I'm using really small toy data, the maximum observed channel is often smaller than total number I want to fake up (maybe I'm using only one spot, for example), and this broke things, so I added these parameters to enable more general functionality like this.